### PR TITLE
Fix comparison of dictionary arrays

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1605

See ticket for more details

### CODE REVIEW

File: datafusion/tests/sql/select.rs [hunk 1]
```diff
@@ -693,8 +705,44 @@ async fn query_on_string_dictionary() -> Result<()> {
...
+        "+-------+",
+        "| d1    |",
+        "+-------+",
+        "| three |",
+        "+-------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    // comparison with another dictionary column
+    let sql = "SELECT d1 FROM test WHERE d1 = d2";
```
Comment (alamb) [inline datafusion/tests/sql/select.rs:721]: this test fails without the change

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 2]
```diff
@@ -77,7 +77,7 @@ pub(crate) fn coerce_types(
 }
 
 fn comparison_eq_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
-    if lhs_type == rhs_type {
+    if lhs_type == rhs_type && !is_dictionary(lhs_type) {
```
Comment (alamb) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:80]: the arrow `eq` kernels don't support `DictionaryArray` <==> `DictionaryArray` comparisons yet, so we need to handle this case specially. I will also file a ticket to add native DictionaryArray comparison support.

File: datafusion/src/physical_plan/expressions/coercion.rs [hunk 3]
```diff
@@ -147,6 +147,10 @@ pub fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<Dat
     }
 }
 
+pub(crate) fn is_dictionary(t: &DataType) -> bool {
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/coercion.rs:150]: The duplication of code between this module and `binary_rule.rs` is not good -- I am going to try and consolidate it  as a follow on PR cc @liukun4515

Comment (alamb) [reply datafusion/src/physical_plan/expressions/coercion.rs:150]: https://github.com/apache/arrow-datafusion/pull/1607 <-- doesn't remove duplication but it at least consolidates the logic into a single module

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs (hunk 2 above)
Comment (liukun4515) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:80]: Can you add `TODO` or some comments for this change?

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs [hunk 4]
```diff
@@ -90,7 +90,7 @@ fn comparison_order_coercion(
     lhs_type: &DataType,
     rhs_type: &DataType,
 ) -> Option<DataType> {
-    if lhs_type == rhs_type {
+    if lhs_type == rhs_type && !is_dictionary(lhs_type) {
```
Comment (liukun4515) [inline datafusion/src/physical_plan/coercion_rule/binary_rule.rs:93]: It's same with above.

**Review by houqp — APPROVED** [review]

File: datafusion/src/physical_plan/expressions/coercion.rs (hunk 3 above)
Comment (liukun4515) [inline datafusion/src/physical_plan/expressions/coercion.rs:150]: another point I want to point out is that the arrow-rs has some codes to check data type, and the logic is also in datafusion, for example, https://docs.rs/arrow/7.0.0/arrow/datatypes/enum.DataType.html#method.is_numeric, Maybe we should refactor them and make the codebase clear.

File: datafusion/src/physical_plan/coercion_rule/binary_rule.rs (hunk 2 above)
Comment (alamb) [reply datafusion/src/physical_plan/coercion_rule/binary_rule.rs:80]: I will do so -- it is a good point

Comment (alamb) [reply datafusion/src/physical_plan/coercion_rule/binary_rule.rs:80]: https://github.com/apache/arrow-rs/issues/1201

Comment (alamb) [reply datafusion/src/physical_plan/coercion_rule/binary_rule.rs:82]: in e27110fd1

File: datafusion/src/physical_plan/expressions/coercion.rs (hunk 3 above)
Comment (alamb) [reply datafusion/src/physical_plan/expressions/coercion.rs:150]: Filed as https://github.com/apache/arrow-datafusion/issues/1613

**Review by liukun4515 — APPROVED** [review]
LGTM
Thanks @alamb

**Pull request merged**